gpflow/utilities/ops.py

```diff
@@ -103,6 +103,29 @@ def square_distance(X, X2):
     return dist
 
 
+def difference_matrix(X, X2):
+    """
+    Returns (X - X2ᵀ)
+
+    This function can deal with leading dimensions in X and X2.
+    For example, If X has shape [M, D] and X2 has shape [N, D],
+    the output will have shape [M, N, D]. If X has shape [I, J, M, D]
+    and X2 has shape [K, L, N, D], the output will have shape
+    [I, J, M, K, L, N, D].
+    """
+    if X2 is None:
+        X2 = X
+        diff = X[..., :, tf.newaxis, :] - X2[..., tf.newaxis, :, :]
+        return diff
+    Xshape = X.shape
+    X2shape = X2.shape
```

Comment:
Actually, this needs to be
```suggestion
    Xshape = tf.shape(X)
    X2shape = tf.shape(X2)
```
to work inside a tf.function() with potentially unknown static tensor shapes.